### PR TITLE
community[minor]: Add CloudBlobLoader that supports loading data from cloud buckets

### PR DESCRIPTION
**Pull request by pprados**
Thank you for contributing to LangChain!

- [ ] **PR title**: "Add CloudBlobLoader"
  - community: Add CloudBlobLoader

- [ ] **PR message**: Add cloud blob loader
    - **Description:** 
 Langchain provides several approaches to read different file formats:

      Specific loaders (`CVSLoader`) or blob-compatible loaders (`FileSystemBlobLoader`). The only implementation proposed for BlobLoader is `FileSystemBlobLoader`.
      
      Many projects retrieve files from cloud storage. We propose a new implementation of `BlobLoader` to read files from the three cloud storage systems. The interface is strictly identical to `FileSystemBlobLoader`. The only difference is the constructor, which takes a cloud "url" object such as  `s3://my-bucket`, `az://my-bucket`, or `gs://my-bucket`.
      
      By streamlining the process, this novel implementation eliminates the requirement to pre-download files from cloud storage to local temporary files (which are seldom removed).
      
      The code relies on the [CloudPathLib](https://cloudpathlib.drivendata.org/stable/) library to interpret cloud URLs. This has been added as an optional dependency.

```Python
loader = CloudBlobLoader("s3://mybucket/id")
for blob in loader.yield_blobs():
    print(blob)
```

- [X] **Dependencies:** CloudPathLib
- [X] **Twitter handle:** pprados

- [X] **Add tests and docs**: Add unit test, but it's easy to convert to integration test, with some files in a cloud storage (see `test_cloud_blob_loader.py`)

- [X] **Lint and test**: Run `make format`, `make lint` and `make test` from the root of the package(s) you've modified.

Hello from Paris @hwchase17. Can you review this PR?

### CODE REVIEW

File: libs/community/langchain_community/chat_models/mlx.py [hunk 1]
```diff
@@ -141,9 +141,6 @@ def _stream(
         run_manager: Optional[CallbackManagerForLLMRun] = None,
         **kwargs: Any,
     ) -> Iterator[ChatGenerationChunk]:
-        import mlx.core as mx
```
Comment (eyurtsev) [inline libs/community/langchain_community/chat_models/mlx.py:144]: unrelated change

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+            Blob instance
+        """
+        if mime_type is None and guess_type:
+            _mimetype = mimetypes.guess_type(path)[0] if guess_type else None  # type: ignore
+        else:
+            _mimetype = mime_type
+        # We do not load the data immediately, instead we treat the blob as a
+        # reference to the underlying data.
+        # data = path.read_bytes()
+        # with open(path, "r", encoding=encoding) as f:
```
Comment (eyurtsev) [inline libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py:271]: Could you delete commented out code?

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+        #     data= f.read_bytes()
+
+        url_parsed = urlparse(str(path))
+        if url_parsed.scheme in ["file", ""]:
```
Comment (eyurtsev) [inline libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py:275]: Why is `file` allowed?

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+        """Count files that match the pattern without loading them."""
+        # Carry out a full iteration to count the files without
+        # materializing anything expensive in memory.
+        num = 0
+        for _ in self._yield_paths():
+            num += 1
+        return num
+
+    @classmethod
+    def from_cloud_path(
```
Comment (eyurtsev) [inline libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py:242]: * Could you replace with static method? 
* Could you rename to `from_path`? (Since it looks like file system is in fact supported, and document which kinds of paths are supported)
* I am assuming this has been attached as a method on the class to avoid exposing the overridden CloudBlob to end users?

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+        """Yield paths that match the requested pattern."""
+        if self.path.is_file():  # type: ignore
+            yield self.path
+            return
+
+        paths = self.path.glob(self.glob)  # type: ignore
+        for path in paths:
+            if self.exclude:
+                if any(path.match(glob) for glob in self.exclude):
+                    continue  # FIXME
```
Comment (eyurtsev) [inline libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py:226]: fix me?

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+            length_func=self.count_matching_files, show_progress=self.show_progress
+        )
+
+        for path in iterator(self._yield_paths()):
+            # yield Blob.from_path(path)
+            yield self.from_cloud_path(path)
+
+    def _yield_paths(self) -> Iterable["AnyPath"]:
+        """Yield paths that match the requested pattern."""
+        if self.path.is_file():  # type: ignore
```
Comment (eyurtsev) [inline libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py:218]: Why is there a type: ignore?

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+            yield self.path
+            return
+
+        paths = self.path.glob(self.glob)  # type: ignore
```
Comment (eyurtsev) [inline libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py:222]: Why is there a type: ignore?

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+    else:
+        iterator = iter  # type: ignore
+
+    return iterator
+
+
+# PUBLIC API
+
+
+class CloudBlobLoader(BlobLoader):
```
Comment (eyurtsev) [inline libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py:126]: Is this meant to support the file system or not?

Code seems to do stuff based on file:// scheme, but documentation and name of loader indicate it's cloud only

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+
+from langchain_community.document_loaders.blob_loaders.schema import (
+    Blob,
+    BlobLoader,
+)
+
+T = TypeVar("T")
+
+
+class CloudBlob(Blob):
```
Comment (eyurtsev) [inline libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py:31]: Could you mark this as private so users, only code to the `Blob` interface and not to the `CloudBlob`?

```suggestion
class _CloudBlob(Blob):
```

**Review by eyurtsev — COMMENTED** [review]
Looks pretty good, would you be able to address comments and we can merge?

File: libs/community/langchain_community/chat_models/mlx.py (hunk 1 above)
Comment (pprados) [reply libs/community/langchain_community/chat_models/mlx.py:144]: I remove 2 line, to pass the lint

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py (hunk 4 above)
Comment (pprados) [reply libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py]: done

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py (hunk 3 above)
Comment (pprados) [reply libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py:271]: The scheme file is to help developers test locally, as if they were in the cloud.

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py (hunk 6 above)
Comment (pprados) [reply libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py:218]: AnyPath is not clear for the typing

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py (hunk 2 above)
Comment (pprados) [reply libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py]: done

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py (hunk 8 above)
Comment (pprados) [reply libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py:126]: Yes, the scheme file is to help developers test locally, as if they were in the cloud.

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py (hunk 9 above)
Comment (pprados) [reply libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py]: done

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+            metadata: Metadata to associate with the blob
+
+        Returns:
+            Blob instance
+        """
+        if mime_type is None and guess_type:
+            _mimetype = mimetypes.guess_type(path)[0] if guess_type else None  # type: ignore
+        else:
+            _mimetype = mime_type
+        #     data= f.read_bytes()
```
Comment (eyurtsev) [inline libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py:268]:
```suggestion
```

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+            return
+
+        paths = self.path.glob(self.glob)
+        for path in paths:
+            if self.exclude:
+                if any(path.match(glob) for glob in self.exclude):
+                    continue
+            if path.is_file():
+                if self.suffixes and path.suffix not in self.suffixes:
+                    continue  # FIXME
```
Comment (eyurtsev) [inline libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py:229]: What needs to be fixed?

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+    .. code-block:: python
+
+        loader = CloudBlobLoader("s3://mybucket/id")
+        for blob in loader.yield_blobs():
+            print(blob)
+    """  # noqa: E501
+
+    def __init__(
+        self,
+        url: Union[str, "AnyPath"],
```
Comment (eyurtsev) [inline libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py:140]: @pprados you probably already know this, but if not

you can use 

`from __future__ import annotations`

And I think that will allow you to not worry about the `"` marks around types that are imported inside the TYPE_CHECKING if conditional block

File: libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+    def count_matching_files(self) -> int:
+        """Count files that match the pattern without loading them."""
+        # Carry out a full iteration to count the files without
+        # materializing anything expensive in memory.
+        num = 0
+        for _ in self._yield_paths():
+            num += 1
+        return num
+
+    @classmethod
```
Comment (eyurtsev) [inline libs/community/langchain_community/document_loaders/blob_loaders/cloud_blob_loader.py:241]: staticmethod probably makes more sense here, but not blocking

**Review by eyurtsev — APPROVED** [review]

**Comment by eyurtsev** [comment]
@pprados let me know if you're okay updating the doc-strings to show file access as well, or if you'd like me to do it

**Comment by eyurtsev** [comment]
@pprados merging right now. If you're able to follow up with a PR to explain the # FIXME comment would be great, and update the doc-string to explain that the file system can be read from as well.
